Add support for Quarkus Fast Jar Packaging

[Quarkus 1.12](https://quarkus.io/blog/quarkus-1-12-0-final-released/) was released last month which made Fast-jar packaging as default. Earlier we used to run `*-runner.jar` in previous versions, but now it's replaced by this:
```
java -jar target/quarkus-app/quarkus-run.jar
```
You can read more about it [here](https://github.com/quarkusio/quarkus/wiki/Migration-Guide-1.12). 

Right now users can't use jkube in zero-configuration  mode with Fast Jar. JKube fails with not being able to find `-runner.jar`:
```
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.eclipse.jkube:kubernetes-maven-plugin:1.1.1:build (default-cli) on project rosa-fruits-app: Execution default-cli of goal org.eclipse.jkube:kubernetes-maven-plugin:1.1.1:build failed: Can't find single file with suffix '-runner.jar' in /home/rohaan/work/repos/rosa-fruits-app/target (zero or more than one files found ending with '-runner.jar') ->
```

 We would need to make changes to QuarkusGenerator to handle Fast-jar packaging as well. We just need to build ImageConfiguration like done in this [Dockerfile here](https://github.com/kameshsampath/rosa-fruits-app/blob/5dd38fd7bbe8ec05d76f175a56d1bb383f4062e8/src/main/docker/Dockerfile.jvm#L3)

### How to Reproduce:
- Try Cloning @kameshsampath 's repo https://github.com/kameshsampath/rosa-fruits-app 
- Remove Image XML configuration
- Try running `k8s:build` or `oc:build`

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -32,6 +32,7 @@ Usage:
 * Fix #677: KarafGenerator includes Jolokia port (8778)
 * Fix #682: Update CircleCI image to new version
 * Fix #666: Replace deprecated JsonParser
+* Fix #679: Quarkus package detection improvements
 
 ### 1.2.0 (2021-03-31)
 * Fix #529: `.maven-dockerignore`, `.maven-dockerexclude`, `.maven-dockerinclude` are no longer supported

--- a/jkube-kit/common/src/main/java/org/eclipse/jkube/kit/common/util/JKubeProjectUtil.java
+++ b/jkube-kit/common/src/main/java/org/eclipse/jkube/kit/common/util/JKubeProjectUtil.java
@@ -109,7 +109,8 @@ public class JKubeProjectUtil {
   }
 
     public static Properties getPropertiesWithSystemOverrides(JavaProject project) {
-        Properties properties = new Properties(project.getProperties());
+        final Properties properties = new Properties();
+        properties.putAll(project.getProperties());
         properties.putAll(System.getProperties());
         return properties;
     }

--- a/jkube-kit/jkube-kit-quarkus/src/main/java/org/eclipse/jkube/quarkus/QuarkusMode.java
+++ b/jkube-kit/jkube-kit-quarkus/src/main/java/org/eclipse/jkube/quarkus/QuarkusMode.java
@@ -1,0 +1,81 @@
+/**
+ * Copyright (c) 2019 Red Hat, Inc.
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at:
+ *
+ *     https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *   Red Hat, Inc. - initial API and implementation
+ */
+package org.eclipse.jkube.quarkus;
+
+import java.io.File;
+import java.util.Arrays;
+import java.util.Optional;
+import java.util.Properties;
+
+import org.eclipse.jkube.kit.common.JavaProject;
+import org.eclipse.jkube.quarkus.generator.QuarkusAssemblies;
+
+import static org.eclipse.jkube.quarkus.QuarkusUtils.getQuarkusConfiguration;
+import static org.eclipse.jkube.quarkus.QuarkusUtils.runnerSuffix;
+
+public enum QuarkusMode {
+
+  NATIVE("native", false, QuarkusAssemblies.NATIVE),
+  FAST_JAR("fast-jar", false, QuarkusAssemblies.FAST_JAR),
+  LEGACY_JAR("legacy-jar", false, QuarkusAssemblies.LEGACY_JAR),
+  UBER_JAR("uber-jar", true, QuarkusAssemblies.UBER_JAR);
+
+  private final String packageType;
+  private final boolean isFatJar;
+  private final QuarkusAssemblies.QuarkusAssembly assembly;
+
+  QuarkusMode(String packageType, boolean isFatJar, QuarkusAssemblies.QuarkusAssembly assembly) {
+    this.packageType = packageType;
+    this.isFatJar = isFatJar;
+    this.assembly = assembly;
+  }
+
+  public boolean isFatJar() {
+    return isFatJar;
+  }
+
+  public QuarkusAssemblies.QuarkusAssembly getAssembly() {
+    return assembly;
+  }
+
+  public static QuarkusMode from(JavaProject project) {
+    final Properties quarkusConfiguration = getQuarkusConfiguration(project);
+    return from(quarkusConfiguration).orElseGet(() -> fromFiles(project, quarkusConfiguration));
+  }
+
+  private static Optional<QuarkusMode> from(Properties properties) {
+    return Optional.ofNullable(properties.getProperty("quarkus.package.type"))
+        .flatMap(packageType -> Arrays.stream(QuarkusMode.values())
+            .filter(qm -> qm.packageType.equalsIgnoreCase(packageType))
+            .findFirst());
+  }
+
+  private static QuarkusMode fromFiles(JavaProject project, Properties properties) {
+    final String runnerSuffix = runnerSuffix(properties);
+    if (hasFileThatEndsWith(project, runnerSuffix + ".jar") && new File(project.getBuildDirectory(), "lib").exists()) {
+      return QuarkusMode.LEGACY_JAR;
+    } else if(hasFileThatEndsWith(project, runnerSuffix + ".jar")) {
+      return QuarkusMode.UBER_JAR;
+    } else if(hasFileThatEndsWith(project, runnerSuffix)) {
+      return QuarkusMode.NATIVE;
+    }
+    return QuarkusMode.FAST_JAR;
+  }
+
+  private static boolean hasFileThatEndsWith(JavaProject project, String suffix) {
+    File buildDir = project.getBuildDirectory();
+    String[] file = buildDir.list((dir, name) -> name.endsWith(suffix));
+    return file != null && file.length > 0;
+  }
+}

--- a/jkube-kit/jkube-kit-quarkus/src/main/java/org/eclipse/jkube/quarkus/QuarkusUtils.java
+++ b/jkube-kit/jkube-kit-quarkus/src/main/java/org/eclipse/jkube/quarkus/QuarkusUtils.java
@@ -16,17 +16,21 @@ package org.eclipse.jkube.quarkus;
 import org.apache.commons.lang3.StringUtils;
 import org.eclipse.jkube.kit.common.JavaProject;
 
+import java.io.File;
 import java.net.URLClassLoader;
 import java.util.Optional;
 import java.util.Properties;
 import java.util.function.Supplier;
 
+import static org.eclipse.jkube.kit.common.util.JKubeProjectUtil.getClassLoader;
 import static org.eclipse.jkube.kit.common.util.PropertiesUtil.getPropertiesFromResource;
+import static org.eclipse.jkube.kit.common.util.PropertiesUtil.toMap;
 import static org.eclipse.jkube.kit.common.util.YamlUtil.getPropertiesFromYamlResource;
 
 public class QuarkusUtils {
 
   private static final String QUARKUS_HTTP_PORT = "quarkus.http.port";
+  private static final String QUARKUS_PACKAGE_RUNNER_SUFFIX = "quarkus.package.runner-suffix";
 
   private QuarkusUtils() {}
 
@@ -41,19 +45,48 @@ public class QuarkusUtils {
     return properties.getProperty(QUARKUS_HTTP_PORT, defaultValue);
   }
 
+  public static String runnerSuffix(Properties properties) {
+    return properties.getProperty(QUARKUS_PACKAGE_RUNNER_SUFFIX, "-runner");
+  }
+
+  public static String findSingleFileThatEndsWith(JavaProject project, String suffix) {
+    final File buildDir = project.getBuildDirectory();
+    String[] file = buildDir.list((dir, name) -> name.endsWith(suffix));
+    if (file == null || file.length != 1) {
+      throw new IllegalStateException("Can't find single file with suffix '" + suffix +
+          "' in " + buildDir + " (zero or more than one files found ending with '" +
+          suffix + "')");
+    }
+    return file[0];
+  }
+
+  /**
+   * Returns the applicable Quarkus configuration properties.
+   *
+   * <p> It takes into account properties provided either via application.properties, or application.yaml files.
+   *
+   * <p> Any property overridden in the project's properties (e.g. Maven properties) will take precedence over the ones
+   * defined in the static configuration files.
+   *
+   * @param project from which to read the configuration
+   * @return the applicable Quarkus configuration properties
+   */
   @SuppressWarnings("unchecked")
-  public static Properties getQuarkusConfiguration(URLClassLoader urlClassLoader) {
+  public static Properties getQuarkusConfiguration(JavaProject project) {
+    final URLClassLoader urlClassLoader = getClassLoader(project);
     final Supplier<Properties>[] sources = new Supplier[]{
         () -> getPropertiesFromResource(urlClassLoader.findResource("application.properties")),
+        () -> getPropertiesFromYamlResource(urlClassLoader.findResource("application.yaml")),
         () -> getPropertiesFromYamlResource(urlClassLoader.findResource("application.yml"))
     };
     for (Supplier<Properties> source : sources) {
       final Properties props = source.get();
       if (!props.isEmpty()) {
+        props.putAll(toMap(project.getProperties()));
         return props;
       }
     }
-    return new Properties();
+    return project.getProperties();
   }
 
   private static Optional<String> getActiveProfile(JavaProject project) {

--- a/jkube-kit/jkube-kit-quarkus/src/main/java/org/eclipse/jkube/quarkus/generator/QuarkusAssemblies.java
+++ b/jkube-kit/jkube-kit-quarkus/src/main/java/org/eclipse/jkube/quarkus/generator/QuarkusAssemblies.java
@@ -1,0 +1,105 @@
+/**
+ * Copyright (c) 2019 Red Hat, Inc.
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at:
+ *
+ *     https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *   Red Hat, Inc. - initial API and implementation
+ */
+package org.eclipse.jkube.quarkus.generator;
+
+import org.eclipse.jkube.kit.common.Assembly;
+import org.eclipse.jkube.kit.common.AssemblyConfiguration;
+import org.eclipse.jkube.kit.common.AssemblyFileSet;
+import org.eclipse.jkube.kit.common.JavaProject;
+import org.eclipse.jkube.kit.common.util.FileUtil;
+import org.eclipse.jkube.kit.common.util.JKubeProjectUtil;
+
+import java.io.File;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Properties;
+
+import static org.eclipse.jkube.quarkus.QuarkusUtils.findSingleFileThatEndsWith;
+import static org.eclipse.jkube.quarkus.QuarkusUtils.getQuarkusConfiguration;
+import static org.eclipse.jkube.quarkus.QuarkusUtils.runnerSuffix;
+
+public class QuarkusAssemblies {
+
+  public static final QuarkusAssembly NATIVE = quarkusGenerator -> {
+    final JavaProject project = quarkusGenerator.getContext().getProject();
+    final Properties quarkusConfiguration = getQuarkusConfiguration(project);
+    final List<String> relativePaths = new ArrayList<>();
+    relativePaths.add(findSingleFileThatEndsWith(project, runnerSuffix(quarkusConfiguration)));
+    final AssemblyFileSet fileSet = AssemblyFileSet.builder()
+        .directory(FileUtil.getRelativePath(project.getBaseDirectory(), project.getBuildDirectory()))
+        .includes(relativePaths)
+        .fileMode("0755")
+        .build();
+    return createAssemblyConfiguration(quarkusGenerator.getBuildWorkdir(), fileSet);
+  };
+
+  public static final QuarkusAssembly FAST_JAR = quarkusGenerator -> {
+    final JavaProject project = quarkusGenerator.getContext().getProject();
+    final File quarkusAppDirectory = new File(project.getBuildDirectory(), "quarkus-app");
+    if (!quarkusAppDirectory.exists()) {
+      throw new IllegalStateException("The quarkus-app directory required in Quarkus Fast Jar mode was not found");
+    }
+    AssemblyFileSet.AssemblyFileSetBuilder fileSetBuilder = AssemblyFileSet.builder()
+        .directory(FileUtil.getRelativePath(project.getBaseDirectory(), quarkusAppDirectory))
+        .include("quarkus-run.jar")
+        .include("*")
+        .include("**/*")
+        .fileMode("0640");
+    addDefaultArtifactExclude(project, fileSetBuilder);
+    return createAssemblyConfiguration(quarkusGenerator.getBuildWorkdir(), fileSetBuilder.build());
+  };
+
+  public static final QuarkusAssembly LEGACY_JAR = quarkusGenerator -> {
+    final JavaProject project = quarkusGenerator.getContext().getProject();
+    AssemblyFileSet.AssemblyFileSetBuilder fileSetBuilder = AssemblyFileSet.builder()
+        .directory(FileUtil.getRelativePath(project.getBaseDirectory(), project.getBuildDirectory()))
+        .include(findSingleFileThatEndsWith(project, runnerSuffix(getQuarkusConfiguration(project)) + ".jar"))
+        .include("lib")
+        .fileMode("0640");
+    addDefaultArtifactExclude(project, fileSetBuilder);
+    return createAssemblyConfiguration(quarkusGenerator.getBuildWorkdir(), fileSetBuilder.build());
+  };
+
+  public static final QuarkusAssembly UBER_JAR = quarkusGenerator -> {
+    final JavaProject project = quarkusGenerator.getContext().getProject();
+    AssemblyFileSet.AssemblyFileSetBuilder fileSetBuilder = AssemblyFileSet.builder()
+        .directory(FileUtil.getRelativePath(project.getBaseDirectory(), project.getBuildDirectory()))
+        .include(findSingleFileThatEndsWith(project, runnerSuffix(getQuarkusConfiguration(project)) + ".jar"))
+        .fileMode("0640");
+    addDefaultArtifactExclude(project, fileSetBuilder);
+    return createAssemblyConfiguration(quarkusGenerator.getBuildWorkdir(), fileSetBuilder.build());
+  };
+
+  private static void addDefaultArtifactExclude(JavaProject project, AssemblyFileSet.AssemblyFileSetBuilder fileSetBuilder) {
+    // We also need to exclude default jar file
+    File defaultJarFile = JKubeProjectUtil.getFinalOutputArtifact(project);
+    if (defaultJarFile != null) {
+      fileSetBuilder.exclude(defaultJarFile.getName());
+    }
+  }
+
+  private static AssemblyConfiguration createAssemblyConfiguration(String targetDir, AssemblyFileSet jKubeAssemblyFileSet) {
+    jKubeAssemblyFileSet.setOutputDirectory(".");
+    return AssemblyConfiguration.builder()
+        .targetDir(targetDir)
+        .excludeFinalOutputArtifact(true)
+        .inline(Assembly.builder().fileSet(jKubeAssemblyFileSet).build())
+        .build();
+  }
+
+  @FunctionalInterface
+  public interface QuarkusAssembly {
+    AssemblyConfiguration createAssemblyConfiguration(QuarkusGenerator quarkusGenerator);
+  }
+}

--- a/jkube-kit/jkube-kit-quarkus/src/main/java/org/eclipse/jkube/quarkus/generator/QuarkusGenerator.java
+++ b/jkube-kit/jkube-kit-quarkus/src/main/java/org/eclipse/jkube/quarkus/generator/QuarkusGenerator.java
@@ -13,7 +13,6 @@
  */
 package org.eclipse.jkube.quarkus.generator;
 
-import java.io.File;
 import java.util.ArrayList;
 import java.util.Collections;
 import java.util.HashMap;
@@ -23,30 +22,25 @@ import java.util.Optional;
 
 import org.eclipse.jkube.generator.api.GeneratorContext;
 import org.eclipse.jkube.generator.javaexec.JavaExecGenerator;
-import org.eclipse.jkube.kit.common.Assembly;
 import org.eclipse.jkube.kit.common.AssemblyConfiguration;
-import org.eclipse.jkube.kit.common.AssemblyFileSet;
 import org.eclipse.jkube.kit.common.Configs;
-import org.eclipse.jkube.kit.common.util.FileUtil;
 import org.eclipse.jkube.kit.common.util.JKubeProjectUtil;
 import org.eclipse.jkube.kit.config.image.ImageConfiguration;
 import org.eclipse.jkube.kit.config.image.build.Arguments;
 import org.eclipse.jkube.kit.config.resource.RuntimeMode;
+import org.eclipse.jkube.quarkus.QuarkusMode;
 
 import lombok.AllArgsConstructor;
 import lombok.Getter;
 import org.apache.commons.lang3.StringUtils;
 
-import static org.eclipse.jkube.kit.common.util.JKubeProjectUtil.getClassLoader;
 import static org.eclipse.jkube.quarkus.QuarkusUtils.extractPort;
+import static org.eclipse.jkube.quarkus.QuarkusUtils.findSingleFileThatEndsWith;
 import static org.eclipse.jkube.quarkus.QuarkusUtils.getQuarkusConfiguration;
+import static org.eclipse.jkube.quarkus.QuarkusUtils.runnerSuffix;
 
 public class QuarkusGenerator extends JavaExecGenerator {
 
-  //For more info on packaging refer https://quarkus.io/guides/getting-started
-  static final String QUARKUS_PACKAGING = "quarkus.package.type";
-  static final String UBER_JAR_PACKAGING = "uber-jar";
-  static final String LEGACY_JAR_PACKAGING = "legacy-jar";
   public static final String QUARKUS = "quarkus";
 
   public QuarkusGenerator(GeneratorContext context) {
@@ -56,7 +50,11 @@ public class QuarkusGenerator extends JavaExecGenerator {
   @AllArgsConstructor
   public enum Config implements Configs.Config {
 
-    // Whether to add native image or plain java image
+    /**
+     * Whether to add native image or plain java image
+     * @deprecated no longer necessary, inferred from Quarkus properties
+     */
+    @Deprecated
     NATIVE_IMAGE("nativeImage", "false");
 
     @Getter
@@ -76,7 +74,7 @@ public class QuarkusGenerator extends JavaExecGenerator {
     final List<String> ports = new ArrayList<>();
     final String quarkusPort = extractPort(
         getProject(),
-        getQuarkusConfiguration(getClassLoader(getProject())),
+        getQuarkusConfiguration(getProject()),
         null);
     addPortIfValid(ports, getConfig(JavaExecGenerator.Config.WEB_PORT, quarkusPort));
     if (!isNativeImage()) {
@@ -97,14 +95,9 @@ public class QuarkusGenerator extends JavaExecGenerator {
   @Override
   protected AssemblyConfiguration createAssembly() {
     if (isNativeImage()) {
-      return createAssemblyConfiguration("/", getNativeFileToInclude());
-    } else if (isUberJar(getQuarkusPackaging())) {
-      return createAssemblyConfiguration(getConfig(JavaExecGenerator.Config.TARGET_DIR), getFatJarFilesToInclude());
-    } else if (isLegacyJar(getQuarkusPackaging())) {
-      return createAssemblyConfiguration(getConfig(JavaExecGenerator.Config.TARGET_DIR), getLegacyJarFilesToInclude());
+      return QuarkusMode.NATIVE.getAssembly().createAssemblyConfiguration(this);
     }
-    return createAssemblyConfiguration(getConfig(JavaExecGenerator.Config.TARGET_DIR),
-        getFastJarFilesToInclude());
+    return QuarkusMode.from(getProject()).getAssembly().createAssemblyConfiguration(this);
   }
 
   @Override
@@ -119,7 +112,7 @@ public class QuarkusGenerator extends JavaExecGenerator {
   protected Arguments getBuildEntryPoint() {
     if (isNativeImage()) {
       final Arguments.ArgumentsBuilder ab = Arguments.builder();
-      ab.execArgument("./" + findSingleFileThatEndsWith(null, "-runner"));
+      ab.execArgument("./" + findSingleFileThatEndsWith(getProject(), runnerSuffix(getQuarkusConfiguration(getProject()))));
       getExtraJavaOptions().forEach(ab::execArgument);
       return ab.build();
     }
@@ -138,7 +131,7 @@ public class QuarkusGenerator extends JavaExecGenerator {
   }
 
   private boolean isNativeImage() {
-    return Boolean.parseBoolean(getConfig(Config.NATIVE_IMAGE));
+    return Boolean.parseBoolean(getConfig(Config.NATIVE_IMAGE)) || QuarkusMode.from(getProject()) == QuarkusMode.NATIVE;
   }
 
   private String getNativeFrom() {
@@ -148,117 +141,9 @@ public class QuarkusGenerator extends JavaExecGenerator {
     return "quay.io/quarkus/ubi-quarkus-native-binary-s2i:1.0";
   }
 
-  private AssemblyConfiguration createAssemblyConfiguration(String targetDir, AssemblyFileSet jKubeAssemblyFileSet) {
-    jKubeAssemblyFileSet.setOutputDirectory(".");
-    return AssemblyConfiguration.builder()
-        .targetDir(targetDir)
-        .excludeFinalOutputArtifact(true)
-        .inline(Assembly.builder().fileSet(jKubeAssemblyFileSet).build())
-        .build();
-  }
-
-  private AssemblyFileSet getLegacyJarFilesToInclude() {
-    AssemblyFileSet.AssemblyFileSetBuilder fileSetBuilder =
-        getFileSetWithFileFromBuildThatEndsWith("-runner.jar");
-    fileSetBuilder.include("lib");
-    // We also need to exclude default jar file
-    File defaultJarFile = JKubeProjectUtil.getFinalOutputArtifact(getContext().getProject());
-    if (defaultJarFile != null) {
-      fileSetBuilder.exclude(defaultJarFile.getName());
-    }
-    fileSetBuilder.fileMode("0640");
-    return fileSetBuilder.build();
-  }
-
-  private AssemblyFileSet getFastJarFilesToInclude() {
-    AssemblyFileSet.AssemblyFileSetBuilder fileSetBuilder = getQuarkusAppDirectory();
-    fileSetBuilder.include("lib").outputDirectory(new File("lib"));
-    fileSetBuilder.include("app").outputDirectory(new File("app"));
-    fileSetBuilder.include(QUARKUS).outputDirectory(new File(QUARKUS));
-    fileSetBuilder.fileMode("0640");
-    return fileSetBuilder.build();
-  }
-
-  private AssemblyFileSet getFatJarFilesToInclude() {
-    AssemblyFileSet.AssemblyFileSetBuilder fileSetBuilder =
-        getFileSetWithFileFromBuildThatEndsWith("-runner.jar");
-    // We also need to exclude default jar file
-    File defaultJarFile = JKubeProjectUtil.getFinalOutputArtifact(getContext().getProject());
-    if (defaultJarFile != null) {
-      fileSetBuilder.exclude(defaultJarFile.getName());
-    }
-    fileSetBuilder.fileMode("0640");
-    return fileSetBuilder.build();
-  }
-
-  private AssemblyFileSet getNativeFileToInclude() {
-    return getFileSetWithFileFromBuildThatEndsWith("-runner")
-        .fileMode("0755")
-        .build();
-  }
-
-  private AssemblyFileSet.AssemblyFileSetBuilder getFileSetWithFileFromBuildThatEndsWith(String suffix) {
-    List<String> relativePaths = new ArrayList<>();
-
-    String fileToInclude = findSingleFileThatEndsWith(null, suffix);
-    if (fileToInclude != null && !fileToInclude.isEmpty()) {
-      relativePaths.add(fileToInclude);
-    }
-    return AssemblyFileSet.builder()
-        .directory(FileUtil.getRelativePath(getProject().getBaseDirectory(), getProject().getBuildDirectory()))
-        .includes(relativePaths)
-        .fileMode("0777");
-  }
-
-  private String findSingleFileThatEndsWith(File buildDir, String suffix) {
-    if(buildDir == null) {
-      buildDir = getProject().getBuildDirectory();
-    }
-    String[] file = buildDir.list((dir, name) -> name.endsWith(suffix));
-    if (file == null || file.length != 1) {
-      throw new IllegalStateException("Can't find single file with suffix '" + suffix +
-          "' in " + buildDir + " (zero or more than one files found ending with '" +
-          suffix + "')");
-    }
-    return file[0];
-  }
-
-  private AssemblyFileSet.AssemblyFileSetBuilder getQuarkusAppDirectory() {
-    List<String> relativePaths = new ArrayList<>();
-    final File projectBuildDir = getProject().getBaseDirectory();
-    final File quarkusAppDir = new File(getProject().getBuildDirectory(), "quarkus-app");
-
-    String fileToInclude = findSingleFileThatEndsWith(quarkusAppDir,"quarkus-run.jar");
-    if (fileToInclude != null && !fileToInclude.isEmpty()) {
-      relativePaths.add(fileToInclude);
-    }
-
-    return AssemblyFileSet.builder()
-        .directory(FileUtil.getRelativePath(projectBuildDir, quarkusAppDir))
-        .includes(relativePaths)
-        .fileMode("0777");
-  }
-
-  private String getQuarkusPackaging() {
-    return JKubeProjectUtil.getPropertiesWithSystemOverrides(getProject())
-        .getProperty(QUARKUS_PACKAGING);
-  }
-
-  private boolean isFastJAR() {
-    return new File(getProject().getBuildDirectory() +
-        "/quarkus-app").exists();
-  }
   @Override
   protected boolean isFatJar() {
-    return isUberJar(getQuarkusPackaging());
-  }
-
-  protected boolean isUberJar(String packaging) {
-    return UBER_JAR_PACKAGING.equals(packaging);
-  }
-
-  private boolean isLegacyJar(String packaging) {
-    return LEGACY_JAR_PACKAGING.equals(packaging) || !isFastJAR();
+    return QuarkusMode.from(getProject()).isFatJar();
   }
 
 }

--- a/jkube-kit/jkube-kit-quarkus/src/test/java/org/eclipse/jkube/quarkus/QuarkusModeTest.java
+++ b/jkube-kit/jkube-kit-quarkus/src/test/java/org/eclipse/jkube/quarkus/QuarkusModeTest.java
@@ -1,0 +1,170 @@
+/**
+ * Copyright (c) 2019 Red Hat, Inc.
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at:
+ *
+ *     https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *   Red Hat, Inc. - initial API and implementation
+ */
+package org.eclipse.jkube.quarkus;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Properties;
+
+import org.eclipse.jkube.kit.common.JavaProject;
+
+import mockit.Expectations;
+import mockit.Mocked;
+import org.apache.commons.io.FileUtils;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.TemporaryFolder;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+@SuppressWarnings("ResultOfMethodCallIgnored")
+public class QuarkusModeTest {
+
+  @Rule
+  public TemporaryFolder temporaryFolder = new TemporaryFolder();
+
+  @Mocked
+  private JavaProject project;
+
+  private File target;
+  private List<String> compileClassPathElements;
+  private Properties projectProperties;
+
+  @Before
+  public void setUp() throws IOException {
+    target = temporaryFolder.newFolder("target");
+    compileClassPathElements = new ArrayList<>();
+    projectProperties = new Properties();
+    // @formatter:off
+    new Expectations() {{
+      project.getCompileClassPathElements(); result = compileClassPathElements;
+      project.getOutputDirectory(); result = target;
+      project.getBuildDirectory(); result = target; minTimes = 0;
+      project.getProperties(); result = projectProperties;
+    }};
+    // @formatter:on
+  }
+
+  @Test
+  public void from_withNoSettingsAndNoFiles_shouldReturnFastJar() {
+    // When
+    final QuarkusMode result = QuarkusMode.from(project);
+    // Then
+    assertThat(result).isEqualTo(QuarkusMode.FAST_JAR);
+  }
+
+  @Test
+  public void from_withNoSettingsAndNativeBinary_shouldReturnNative() throws IOException {
+    // Given
+    new File(target, "-runner").createNewFile();
+    // When
+    final QuarkusMode result = QuarkusMode.from(project);
+    // Then
+    assertThat(result).isEqualTo(QuarkusMode.NATIVE);
+  }
+
+  @Test
+  public void from_withCustomSuffixAndNativeBinary_shouldReturnNative() throws IOException {
+    // Given
+    projectProperties.put("quarkus.package.runner-suffix", "-coyote");
+    new File(target, "-coyote").createNewFile();
+    // When
+    final QuarkusMode result = QuarkusMode.from(project);
+    // Then
+    assertThat(result).isEqualTo(QuarkusMode.NATIVE);
+  }
+
+  @Test
+  public void from_withNoSettingsAndRunnerJar_shouldReturnUberJar() throws IOException {
+    // Given
+    new File(target, "-runner.jar").createNewFile();
+    // When
+    final QuarkusMode result = QuarkusMode.from(project);
+    // Then
+    assertThat(result).isEqualTo(QuarkusMode.UBER_JAR);
+  }
+
+  @Test
+  public void from_withCustomSuffixAndRunnerJar_shouldReturnUberJar() throws IOException {
+    // Given
+    projectProperties.put("quarkus.package.runner-suffix", "-coyote");
+    new File(target, "-coyote.jar").createNewFile();
+    // When
+    final QuarkusMode result = QuarkusMode.from(project);
+    // Then
+    assertThat(result).isEqualTo(QuarkusMode.UBER_JAR);
+  }
+
+  @Test
+  public void from_withNoSettingsAndRunnerJarAndLib_shouldReturnLegacyJar() throws IOException {
+    // Given
+    new File(target, "-runner.jar").createNewFile();
+    FileUtils.forceMkdir(new File(target, "lib"));
+    // When
+    final QuarkusMode result = QuarkusMode.from(project);
+    // Then
+    assertThat(result).isEqualTo(QuarkusMode.LEGACY_JAR);
+  }
+
+  @Test
+  public void from_withNativePackaging_shouldReturnNative() {
+    // Given
+    projectProperties.put("quarkus.package.type", "native");
+    // When
+    final QuarkusMode result = QuarkusMode.from(project);
+    // Then
+    assertThat(result)
+        .isEqualTo(QuarkusMode.NATIVE)
+        .hasFieldOrPropertyWithValue("fatJar", false);
+  }
+
+  @Test
+  public void from_withLegacyJarPackaging_shouldReturnLegacyJar() {
+    // Given
+    projectProperties.put("quarkus.package.type", "legacy-jar");
+    // When
+    final QuarkusMode result = QuarkusMode.from(project);
+    // Then
+    assertThat(result)
+        .isEqualTo(QuarkusMode.LEGACY_JAR)
+        .hasFieldOrPropertyWithValue("fatJar", false);
+  }
+
+  @Test
+  public void from_withFastJarPackaging_shouldReturnFastJar() {
+    // Given
+    projectProperties.put("quarkus.package.type", "fast-jar");
+    // When
+    final QuarkusMode result = QuarkusMode.from(project);
+    // Then
+    assertThat(result)
+        .isEqualTo(QuarkusMode.FAST_JAR)
+        .hasFieldOrPropertyWithValue("fatJar", false);
+  }
+
+  @Test
+  public void from_withUberJarPackaging_shouldReturnUberJar() {
+    // Given
+    projectProperties.put("quarkus.package.type", "uber-jar");
+    // When
+    final QuarkusMode result = QuarkusMode.from(project);
+    // Then
+    assertThat(result)
+        .isEqualTo(QuarkusMode.UBER_JAR)
+        .hasFieldOrPropertyWithValue("fatJar", true);
+  }
+}

--- a/jkube-kit/jkube-kit-quarkus/src/test/java/org/eclipse/jkube/quarkus/QuarkusUtilsTest.java
+++ b/jkube-kit/jkube-kit-quarkus/src/test/java/org/eclipse/jkube/quarkus/QuarkusUtilsTest.java
@@ -13,14 +13,17 @@
  */
 package org.eclipse.jkube.quarkus;
 
-import java.net.URL;
-import java.net.URLClassLoader;
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.Collections;
 import java.util.Properties;
 
 import org.eclipse.jkube.kit.common.JavaProject;
 
 import org.junit.Before;
+import org.junit.Rule;
 import org.junit.Test;
+import org.junit.rules.TemporaryFolder;
 
 import static org.assertj.core.api.Assertions.assertThat;
 import static org.assertj.core.api.Assertions.entry;
@@ -29,11 +32,17 @@ import static org.eclipse.jkube.quarkus.QuarkusUtils.getQuarkusConfiguration;
 
 public class QuarkusUtilsTest {
 
+  @Rule
+  public TemporaryFolder temporaryFolder = new TemporaryFolder();
+
   private JavaProject javaProject;
 
   @Before
-  public void setUp() {
-    javaProject = JavaProject.builder().properties(new Properties()).build();
+  public void setUp() throws IOException {
+    javaProject = JavaProject.builder()
+        .properties(new Properties())
+        .outputDirectory(temporaryFolder.newFolder())
+        .build();
   }
 
   @Test
@@ -81,14 +90,30 @@ public class QuarkusUtilsTest {
   }
 
   @Test
+  public void getQuarkusConfiguration_propertiesAndYamlProjectProperties_shouldUseProjectProperties() {
+    // Given
+    javaProject.getProperties().put("quarkus.http.port", "42");
+    javaProject.setCompileClassPathElements(Arrays.asList(
+        QuarkusUtilsTest.class.getResource("/utils-test/config/yaml/").getPath(),
+        QuarkusUtilsTest.class.getResource("/utils-test/config/properties/").getPath()
+    ));
+    // When
+    final Properties props = getQuarkusConfiguration(javaProject);
+    // Then
+    assertThat(props).containsOnly(
+        entry("quarkus.http.port", "42"),
+        entry("%dev.quarkus.http.port", "8082"));
+  }
+
+  @Test
   public void getQuarkusConfiguration_propertiesAndYaml_shouldUseProperties() {
     // Given
-    final URLClassLoader ucl = URLClassLoader.newInstance(new URL[] {
-        QuarkusUtilsTest.class.getResource("/utils-test/config/yaml/"),
-        QuarkusUtilsTest.class.getResource("/utils-test/config/properties/")
-    });
+    javaProject.setCompileClassPathElements(Arrays.asList(
+        QuarkusUtilsTest.class.getResource("/utils-test/config/yaml/").getPath(),
+        QuarkusUtilsTest.class.getResource("/utils-test/config/properties/").getPath()
+    ));
     // When
-    final Properties props = getQuarkusConfiguration(ucl);
+    final Properties props = getQuarkusConfiguration(javaProject);
     // Then
     assertThat(props).containsOnly(
         entry("quarkus.http.port", "1337"),
@@ -98,11 +123,11 @@ public class QuarkusUtilsTest {
   @Test
   public void getQuarkusConfiguration_yamlOnly_shouldUseYaml() {
     // Given
-    final URLClassLoader ucl = URLClassLoader.newInstance(new URL[] {
-        QuarkusUtilsTest.class.getResource("/utils-test/config/yaml/")
-    });
+    javaProject.setCompileClassPathElements(Collections.singletonList(
+        QuarkusUtilsTest.class.getResource("/utils-test/config/yaml/").getPath()
+    ));
     // When
-    final Properties props = getQuarkusConfiguration(ucl);
+    final Properties props = getQuarkusConfiguration(javaProject);
     // Then
     assertThat(props).containsOnly(
         entry("quarkus.http.port", "31337"),
@@ -112,11 +137,11 @@ public class QuarkusUtilsTest {
   @Test
   public void getQuarkusConfiguration_noConfigFiles_shouldReturnEmpty() {
     // Given
-    final URLClassLoader ucl = URLClassLoader.newInstance(new URL[] {
-        QuarkusUtilsTest.class.getResource("/")
-    });
+    javaProject.setCompileClassPathElements(Collections.singletonList(
+        QuarkusUtilsTest.class.getResource("/").getPath()
+    ));
     // When
-    final Properties props = getQuarkusConfiguration(ucl);
+    final Properties props = getQuarkusConfiguration(javaProject);
     // Then
     assertThat(props).isEmpty();
   }

--- a/jkube-kit/jkube-kit-quarkus/src/test/java/org/eclipse/jkube/quarkus/generator/QuarkusGeneratorTest.java
+++ b/jkube-kit/jkube-kit-quarkus/src/test/java/org/eclipse/jkube/quarkus/generator/QuarkusGeneratorTest.java
@@ -20,6 +20,7 @@ import java.util.Collections;
 import java.util.List;
 import java.util.Properties;
 
+import org.apache.commons.io.FileUtils;
 import org.eclipse.jkube.generator.api.DefaultImageLookup;
 import org.eclipse.jkube.generator.api.GeneratorContext;
 import org.eclipse.jkube.kit.common.Assembly;
@@ -39,6 +40,8 @@ import org.junit.Test;
 import org.junit.rules.TemporaryFolder;
 
 import static org.assertj.core.api.Assertions.assertThat;
+import static org.junit.Assert.assertThrows;
+
 /**
  * @author jzuriaga
  */
@@ -70,8 +73,6 @@ public class QuarkusGeneratorTest {
     projectProps = new Properties();
     projectProps.put("jkube.generator.name", "quarkus");
     baseDir = temporaryFolder.newFolder("target");
-    createFakeRunnerJar(false);
-    createFakeNativeImage(false);
     // @formatter:off
     new Expectations() {{
       project.getVersion(); result = "0.0.1-SNAPSHOT"; minTimes = 0;
@@ -182,7 +183,17 @@ public class QuarkusGeneratorTest {
   }
 
   @Test
-  public void assembly_withDefaults_shouldReturnDefaultAssemblyInImage() {
+  public void isFatJar_withDefaults_shouldBeFalse() {
+    // When
+    final boolean result = new QuarkusGenerator(ctx).isFatJar();
+    // Then
+    assertThat(result).isFalse();
+  }
+
+  @Test
+  public void assembly_withFastJarInTarget_shouldReturnFastJarAssemblyInImage() throws IOException {
+    // Given
+    withFastJarInTarget();
     // When
     final List<ImageConfiguration> resultImages = new QuarkusGenerator(ctx)
         .customize(new ArrayList<>(), false);
@@ -199,11 +210,11 @@ public class QuarkusGeneratorTest {
         .asList()
         .hasSize(1)
         .flatExtracting("includes")
-        .containsExactlyInAnyOrder("lib", "sample-runner.jar");
+        .containsExactly("quarkus-run.jar", "*", "**/*");
   }
 
   @Test
-  public void assembly_withDefaultsInNative_shouldReturnDefaultNativeAssemblyInImage() throws IOException {
+  public void assembly_manualNativeSettings_shouldReturnNativeAssemblyInImage() throws IOException {
     // Given
     setNativeConfig(false);
     // When
@@ -226,16 +237,32 @@ public class QuarkusGeneratorTest {
   }
 
   @Test
-  public void isFatJar_withDefaults_shouldBeFalse() {
+  public void assembly_withNativeBinaryInTarget_shouldReturnNativeAssemblyInImage() throws IOException {
+    // Given
+    withNativeBinaryInTarget();
     // When
-    final boolean result = new QuarkusGenerator(ctx).isFatJar();
+    final List<ImageConfiguration> resultImages = new QuarkusGenerator(ctx)
+        .customize(new ArrayList<>(), false);
     // Then
-    assertThat(result).isFalse();
+    assertThat(resultImages)
+        .isNotNull()
+        .hasSize(1)
+        .element(0)
+        .extracting(ImageConfiguration::getBuild)
+        .extracting(BuildConfiguration::getAssembly)
+        .hasFieldOrPropertyWithValue("targetDir", "/")
+        .extracting(AssemblyConfiguration::getInline)
+        .extracting(Assembly::getFileSets)
+        .asList()
+        .hasSize(1)
+        .flatExtracting("includes")
+        .containsExactly("sample-runner");
   }
 
   @Test
-  public void assembly_withLegacyJar_shouldReturnDefaultAssemblyInImage() {
-    projectProps.put("quarkus.package.type","legacy-jar");
+  public void assembly_withLegacyJarInTarget_shouldReturnDefaultAssemblyInImage() throws IOException {
+    // Given
+    withLegacyJarInTarget();
     // When
     final List<ImageConfiguration> resultImages = new QuarkusGenerator(ctx)
         .customize(new ArrayList<>(), false);
@@ -252,13 +279,28 @@ public class QuarkusGeneratorTest {
         .asList()
         .hasSize(1)
         .flatExtracting("includes")
-        .containsOnly("lib","sample-runner.jar");
-    projectProps.remove("quarkus.package.type");
+        .containsOnly("lib", "sample-legacy-runner.jar");
   }
 
   @Test
-  public void assembly_withUberJar_shouldReturnDefaultAssemblyInImage() {
-    projectProps.put("quarkus.package.type","uber-jar");
+  public void assembly_withConfiguredPackagingAndNoJar_shouldThrowException() throws IOException {
+    // Given
+    projectProps.put("quarkus.package.type", "legacy-jar");
+    final QuarkusGenerator quarkusGenerator = new QuarkusGenerator(ctx);
+    final List<ImageConfiguration> list = new ArrayList<>();
+    // When
+    final IllegalStateException result = assertThrows(IllegalStateException.class, () ->
+        quarkusGenerator.customize(list, false));
+    // Then
+    assertThat(result)
+        .hasMessageContaining("Can't find single file with suffix '-runner.jar'");
+  }
+
+  @Test
+  public void assembly_withUberJarInTarget_shouldReturnAssemblyWithSingleJar() throws IOException {
+    // Given
+//    projectProps.put("quarkus.package.type", "uber-jar");
+    withUberJarInTarget();
     // When
     final List<ImageConfiguration> resultImages = new QuarkusGenerator(ctx)
         .customize(new ArrayList<>(), false);
@@ -276,13 +318,12 @@ public class QuarkusGeneratorTest {
         .hasSize(1)
         .flatExtracting("includes")
         .containsExactly("sample-runner.jar");
-    projectProps.remove("quarkus.package.type");
   }
 
   @Test
-  public void assembly_withFastJar_shouldReturnDefaultAssemblyInImage() throws IOException {
+  public void assembly_withFastJarInTarget_shouldReturnAssemblyForQuarkusAppInImage() throws IOException {
     // Given
-    createFakeRunnerJar(true);
+    withFastJarInTarget();
     // When
     final List<ImageConfiguration> resultImages = new QuarkusGenerator(ctx)
         .customize(new ArrayList<>(), false);
@@ -299,7 +340,7 @@ public class QuarkusGeneratorTest {
         .asList()
         .hasSize(1)
         .flatExtracting("includes")
-        .containsExactly("quarkus-run.jar","lib","app", "quarkus");
+        .containsExactly("quarkus-run.jar", "*", "**/*");
   }
 
   private void assertBuildFrom (List<ImageConfiguration> resultImages, String baseImage) {
@@ -310,11 +351,28 @@ public class QuarkusGeneratorTest {
         .containsExactly(baseImage);
   }
 
-  private void createFakeRunnerJar(boolean fastJAR) throws IOException {
-    File runnerJar = new File(
-        fastJAR ? temporaryFolder.newFolder("target/quarkus-app") : baseDir,
-        fastJAR ? "quarkus-run.jar" : "sample-runner.jar");
-    runnerJar.createNewFile();
+  private void withNativeBinaryInTarget() throws IOException {
+    new File(baseDir, "sample-runner").createNewFile();
+  }
+
+  private void withUberJarInTarget() throws IOException {
+    new File(baseDir, "sample-runner.jar").createNewFile();
+  }
+
+  private void withLegacyJarInTarget() throws IOException {
+    new File(baseDir, "sample-legacy-runner.jar").createNewFile();
+    final File lib = new File(baseDir, "lib");
+    FileUtils.forceMkdir(lib);
+    new File(lib, "dependency.jar");
+  }
+
+  private void withFastJarInTarget() throws IOException {
+    final File quarkusApp = new File(baseDir, "quarkus-app");
+    FileUtils.forceMkdir(quarkusApp);
+    FileUtils.forceMkdir(new File(quarkusApp, "app"));
+    FileUtils.forceMkdir(new File(quarkusApp, "lib"));
+    FileUtils.forceMkdir(new File(quarkusApp, "quarkus"));
+    new File(quarkusApp, "quarkus-run.jar").createNewFile();
   }
 
   private void in(RuntimeMode runtimeMode) {

--- a/quickstarts/maven/quarkus/pom.xml
+++ b/quickstarts/maven/quarkus/pom.xml
@@ -32,9 +32,22 @@
     <maven.compiler.source>1.8</maven.compiler.source>
     <maven.compiler.target>1.8</maven.compiler.target>
     <jkube.version>${project.version}</jkube.version>
-    <version.quarkus>1.7.0.Final</version.quarkus>
+    <version.quarkus>1.13.1.Final</version.quarkus>
+    <quarkus.package.type>legacy-jar</quarkus.package.type>
     <jkube.enricher.jkube-service.type>NodePort</jkube.enricher.jkube-service.type>
   </properties>
+
+  <dependencyManagement>
+    <dependencies>
+      <dependency>
+        <groupId>io.quarkus</groupId>
+        <artifactId>quarkus-bom</artifactId>
+        <version>${version.quarkus}</version>
+        <type>pom</type>
+        <scope>import</scope>
+      </dependency>
+    </dependencies>
+  </dependencyManagement>
 
   <dependencies>
     <dependency>

